### PR TITLE
[User Model] Update Notifications API

### PR DESCRIPTION
**Pull request by jennantilla**
# Description
## One Line Summary
Update Notifications API to include new naming and events.

## Details

1. update `set` to `add` for listeners
2. update`open` to `click`
3. support adding multiple notification click listeners, and removal
4. Replace object OSNotificationOpenedResult -> OSNotificationClickEvent
5. Replace object OSNotificationAction -> OSNotificationClickResult
6. Update permission events on Android

### Motivation
Align with native SDK naming conventions and events.

### Scope
Changes limited to Notifications namespace.

# Testing
## Manual testing
Tested running plugin on both Samsung Galaxy S21 and various Android emulators all running Android 13. Tested running plugin on 6th gen iPad running 15.2 and later upgraded to 16.4.1.

# Affected code checklist
   - [X] Notifications
      - [X] Display
      - [X] Open
      - [X] Push Processing
      - [ ] Confirm Deliveries
   - [ ] Outcomes
   - [ ] Sessions
   - [ ] In-App Messaging
   - [ ] REST API requests
   - [X] Public API changes

# Checklist
## Overview
   - [X] I have filled out all **REQUIRED** sections above
   - [X] PR does one thing
     - If it is hard to explain how any codes changes are related to each other then it most likely needs to be more than one PR
   - [x] Any Public API changes are explained in the PR details and conform to existing APIs

## Testing
   - [X] I have included test coverage for these changes, or explained why they are not needed
   - [X] All automated tests pass, or I explained why that is not possible
   - [X] I have personally tested this on my device, or explained why that is not possible

## Final pass
   - [X] Code is as readable as possible.
      - Simplify with less code, followed by splitting up code into well named functions and variables, followed by adding comments to the code.
   - [X] I have reviewed this PR myself, ensuring it meets each checklist item
      - WIP (Work In Progress) is ok, but explain what is still in progress and what you would like feedback on. Start the PR title with "WIP" to indicate this.

- - -
This change is [Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/887)

### CODE REVIEW

File: src/ios/OneSignalPush.m [hunk 1]
```diff
@@ -177,45 +185,48 @@ - (void)setProvidesNotificationSettingsView:(CDVInvokedUrlCommand *)command {
...
-
-    [OneSignal.Notifications setNotificationWillShowInForegroundHandler:^(OSNotification *notification, OSNotificationDisplayResponse completion) {
-        self.receivedNotificationCache[notification.notificationId] = notification;
-        self.notificationCompletionCache[notification.notificationId] = completion;
-        processNotificationWillShowInForeground(notification);
-    }];
+- (void)onWillDisplayNotification:(OSNotificationWillDisplayEvent *)event {
+    self.notificationWillDisplayCache[event.notification.notificationId] = event;
+    processForegroundLifecycleListener(event);
+    [event preventDefault];
```
Comment (emawby) [inline src/ios/OneSignalPush.m:191]: Do we need to call preventDefault before firing the Cordova callback? My concern is if someone calls notification.display synchronously in their callback we would then call preventDefault after that. I don't know if that has any negative affects though.

File: src/ios/OneSignalPush.m [hunk 2]
```diff
@@ -177,45 +185,48 @@ - (void)setProvidesNotificationSettingsView:(CDVInvokedUrlCommand *)command {
...
+    self.preventDefaultCache[event.notification.notificationId] = event;
 }
 
-- (void)completeNotification:(CDVInvokedUrlCommand*)command {
+-(void)displayNotification:(CDVInvokedUrlCommand *)command {
     NSString *notificationId = command.arguments[0];
-    BOOL shouldDisplay = [command.arguments[1] boolValue];
-    OSNotificationDisplayResponse completion = self.notificationCompletionCache[notificationId];
+    OSNotificationWillDisplayEvent *event = _notificationWillDisplayCache[notificationId];
```
Comment (emawby) [inline src/ios/OneSignalPush.m:203]: What happens if event is nil after trying to retrieve it from the cache.

File: www/NotificationsNamespace.ts [hunk 3]
```diff
@@ -1,16 +1,16 @@
...
 import OSNotification from './OSNotification';
 
 // Suppress TS warnings about window.cordova
 declare let window: any; // turn off type checking
 
 export default class Notifications {
     private _permissionObserverList: ((event:boolean)=>void)[] = [];
-    private _notificationOpenedDelegate = function(notificationOpened: OpenedEvent) {};
-    private _notificationWillShowInForegroundDelegate = function(notificationReceived: NotificationReceivedEvent) {};
+    private _notificationClickedListener = [function(notificationClicked: ClickedEvent) {}];
```
Comment (emawby) [inline www/NotificationsNamespace.ts:10]: lets pluralize this field since it is an array

**Review by emawby — CHANGES_REQUESTED** [review]

File: src/ios/OneSignalPush.m (hunk 2 above)
Comment (jennantilla) [reply src/ios/OneSignalPush.m:203]: Oh yeah, we'd want to return:

if (!event ) {
        return;
}

Makes sense to add some logging, too.

File: src/ios/OneSignalPush.m (hunk 1 above)
Comment (jennantilla) [reply src/ios/OneSignalPush.m:191]: Like moving preventDefault before processForegroundLifecycleListener? I'm not sure of any impact either, but I've been experimenting with them swapped and haven't noticed any issues yet.

**Comment by jennantilla** [comment]
@jkasten2 @emawby sorry for the delay and messy commits. This should now be ready for review.

I did not add the notification lifecycle listener on init because I realized that unless the user explicitly calls `addEventListener("foregroundWillDisplay", event {})`, then preventDefault will fire. Definitely let me know if I should have implemented differently to address this!

**Review by emawby — CHANGES_REQUESTED** [review]
> Reviewed 1 of 8 files at r1, 2 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/887)* status: 3 of 8 files reviewed, 1 unresolved discussion (waiting on @jennantilla, @jkasten2, @nan-li, and @shepherd-l)

___
*[`src/android/com/onesignal/cordova/OneSignalPush.java` line 651 at r3](https://reviewable.io/reviews/OneSignal/OneSignal-Cordova-SDK/887#-NW3AjjP4O4EkTAKN_it:-NW3AjjP4O4EkTAKN_iu:b33l4b3) ([raw file](https://github.com/OneSignal/OneSignal-Cordova-SDK/blob/aa2f6e5c1296cf6b698d6f3bb2e9dffc31017da9/src/android/com/onesignal/cordova/OneSignalPush.java#L651)):*
> ```Java
>   @Override
>   public void onDestroy() {
>     OneSignal.getNotifications().addClickListener(null);
> ```

These should be removeClickListener rather than add null

File: src/android/com/onesignal/cordova/OneSignalPush.java [hunk 4]
```diff
@@ -186,6 +262,11 @@ public boolean init(CallbackContext callbackContext, JSONArray data) {
       String appId = data.getString(0);
       OneSignal.initWithContext(this.cordova.getActivity(), appId);
 
+
+      // Notification listener
+      CordovaNotificationClickListener cordovaNotificationClickListener = new CordovaNotificationClickListener(callbackContext);
```
Comment (jkasten2) [inline src/android/com/onesignal/cordova/OneSignalPush.java:267]: Setting this up here will result in click events being missed on app cold start. Unless the app developer called `addClickListener` before initialization. It's something that could technically be documented, however it's much better to not have this restriction as its too easy to overlook.

I recommend setup up the listener only after the app developer calls `addClickListener`. This is simpler than  writing code in this wrapper to store the click event then forwarding on, as the OneSignal native SDK already has this logic.

**Review by jkasten2 — CHANGES_REQUESTED** [review]

**Review by jkasten2 — APPROVED** [review]

**Pull request merged**